### PR TITLE
ISSUES-180 fix empty function with FixedString

### PR DESCRIPTION
**Pull request by zhang2014**
#180
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

### CODE REVIEW

File: dbms/src/Functions/FunctionsString.cpp [hunk 1]
```diff
@@ -42,13 +44,16 @@ struct EmptyImpl
...
-        res = negative ^ (n == 0);
     }
 
-    static void vector_fixed_to_vector(const ColumnString::Chars_t & /*data*/, size_t /*n*/, PaddedPODArray<UInt8> & /*res*/)
+    static void vector_fixed_to_vector(const ColumnString::Chars_t & data, size_t n, PaddedPODArray<UInt8> & res)
     {
+        size_t size = data.size() / n;
+
+        for (size_t i = 0; i < size; ++i)
+            res[i] = negative ^ (n == 0 || data[i * n] == '\0');
```
Comment (alexey-milovidov) [inline dbms/src/Functions/FunctionsString.cpp]: What if the value is '\0abc'?

**Comment by zhang2014** [comment]
@alexey-milovidov Please help review the changes

**Comment by alexey-milovidov** [comment]
Almost perfect :)
Just few points...

File: dbms/src/Functions/FunctionsString.cpp [hunk 2]
```diff
@@ -42,13 +44,17 @@ struct EmptyImpl
         }
     }
 
-    static void vector_fixed_to_constant(const ColumnString::Chars_t & /*data*/, size_t n, UInt8 & res)
+    static void vector_fixed_to_constant(const ColumnString::Chars_t & /*data*/, size_t /*n*/, UInt8 & /*res*/)
```
Comment (alexey-milovidov) [inline dbms/src/Functions/FunctionsString.cpp]: We may throw here with "Logical error" code.

File: dbms/src/Functions/FunctionsString.cpp [hunk 3]
```diff
@@ -31,6 +31,8 @@ using namespace GatherUtils;
 template <bool negative = false>
 struct EmptyImpl
 {
+    static constexpr auto is_fixed_to_constant = false;
```
Comment (alexey-milovidov) [inline dbms/src/Functions/FunctionsString.cpp]: Missing comment.

File: dbms/src/Functions/FunctionsString.cpp [hunk 4]
```diff
@@ -42,13 +44,17 @@ struct EmptyImpl
...
     }
 
-    static void vector_fixed_to_vector(const ColumnString::Chars_t & /*data*/, size_t /*n*/, PaddedPODArray<UInt8> & /*res*/)
+    static void vector_fixed_to_vector(const ColumnString::Chars_t & data, size_t n, PaddedPODArray<UInt8> & res)
     {
+        char * empty_chars[n] = {0};
+        size_t size = data.size() / n;
+
+        for (size_t i = 0; i < size; ++i)
+            res[i] = negative ^ (n == 0 || 0 == memcmp(&data[i * size], empty_chars, n));
```
Comment (alexey-milovidov) [inline dbms/src/Functions/FunctionsString.cpp]: BTW, there is `memoryIsZero` function, but it is optimized only for huge memory arrays.
Your solution is fine.

File: dbms/src/Functions/FunctionsString.cpp [hunk 5]
```diff
@@ -42,13 +44,17 @@ struct EmptyImpl
...
-    static void vector_fixed_to_constant(const ColumnString::Chars_t & /*data*/, size_t n, UInt8 & res)
+    static void vector_fixed_to_constant(const ColumnString::Chars_t & /*data*/, size_t /*n*/, UInt8 & /*res*/)
     {
-        res = negative ^ (n == 0);
     }
 
-    static void vector_fixed_to_vector(const ColumnString::Chars_t & /*data*/, size_t /*n*/, PaddedPODArray<UInt8> & /*res*/)
+    static void vector_fixed_to_vector(const ColumnString::Chars_t & data, size_t n, PaddedPODArray<UInt8> & res)
     {
+        char * empty_chars[n] = {0};
```
Comment (alexey-milovidov) [inline dbms/src/Functions/FunctionsString.cpp]: MAX_FIXEDSTRING_SIZE is 16 MB and this code can lead to stack overflow
(16 MB is impractical size for FixedString, but malicious user can pass this value to kill the server.)

Comment (alexey-milovidov) [reply dbms/src/Functions/FunctionsString.cpp]: And missing `const`.

Comment (alexey-milovidov) [reply dbms/src/Functions/FunctionsString.cpp]: And it doesn't compile with clang 7.

File: dbms/src/Functions/FunctionsString.cpp (hunk 4 above)
Comment (alexey-milovidov) [inline dbms/src/Functions/FunctionsString.cpp]: The check for `n == 0` contradict the fact that we was dividing by n.
And actually `FixedString(0)` is not possible.

Comment (zhang2014) [reply dbms/src/Functions/FunctionsString.cpp]: Thank for your guidance.

**Comment by zhang2014** [comment]
It would be better if I could fix the problem, It helps me grow  : )

**Pull request merged**